implement custom builds, for `dev` and `publish`

threepointone opened the pull request:
The code here is a bit jank, but we should probably refactor the way we pass arguments to dev and publish to 'fix' it properly. Anyway.

This PR adds custom builds to wrangler2 (https://developers.cloudflare.com/workers/cli-wrangler/configuration#build).
- In `wrangler.toml`, you can configure `build.command` and`build.cwd` and  and it'll be called before `publish` or `dev`. We also optionally watch `build.watch_dir` for changes during dev. There's some discussion to be had whether we're going to deprecate this config commands, but we'll support it until then anyway.

I swear I'll start writing tests for `publish` before the year is over. Maybe even `dev`.

closes #120

Electroid commented on packages/wrangler/src/index.tsx:
```diff
@@ -390,7 +390,7 @@ export async function main(argv: string[]): Promise<void> {
   // dev
   yargs.command(
-    "dev <filename>",
+    "dev <filename> [-- build]",
```
Could we add other, non-build, flags in the future that would be compatible with this? e.g. `--dry-run`

threepointone replied:
Yeah for sure. We already support `--port`, we can add more. 

Honestly, I'm not fully sold on this blank `--` thing yet, could be confusing. Might be worth just having explicit `--build <command>`.

petebacondarwin replied:
` -- ` makes the `build` argument become positional. Arguably in this case we should support this positional being used without the `--`. E.g. `wrangler dev npm build`, rather than `wrangler dev -- npm build`. At least this is my understanding of the convention around `--` (i.e. it indicates the end of options and anything after is positional).

So with that in mind, I would prefer to have build as a clear option: `wrangler dev --build "npm build"`.

threepointone replied:
as discussed, I removed the `--` option altogether.

threepointone commented on packages/wrangler/src/dev.tsx:
```diff
@@ -371,7 +424,7 @@ type EsbuildBundle = {
 
 function useEsbuild(props: {
-  entry: string;
+  entry: void | string;
```
I need to revisit very `void` and replace it with `undefined`. I just got trigger happy and regret it now.

threepointone commented:
I made some changes - 
- removed the `--` option completely. This is better accomplished with a bash script, and we should have great documentation, and a terminal log, showing how to do it. We can do that in a separate PR.
- I also added support for `config.build.watch_dir`. We can deprecate it when (and if!) we deprecate `config.build` altogether. 

I'll edit the main PR description as well.

GregBrimble commented:
Any thoughts on `--` in `pages dev`? Feels slightly different to me, but maybe we value a shared interface across wrangler more?

threepointone commented:
Pages dev has an additional requirement of 'detecting' what ports were opened and proxying them, etc, so I can see the requirement there. That said, _I'd_ revisit the choice of the blank `--`, and maybe consider using `--build` instead. That's a product decision for you folks tho, and I assume your users have feedback about it too.

threepointone commented:
(That said, I'd like for `pages dev` to use `wrangler dev` at some point, so I imagine we'll converge on some solution.)

petebacondarwin commented on packages/wrangler/src/dev.tsx:
```diff
@@ -391,13 +397,36 @@ function useCustomBuild(
+      watch(watch_dir, { persistent: true, ignoreInitial: true }).on(
+        "all",
+        (_event, _path) => {
+          console.log("Restarting build...");
```
Perhaps log the path of the changed file that caused the rebuild?

Something like?

```ts
console.log(`The file ${path} changed`);
```

threepointone replied:
done and done

petebacondarwin commented on packages/wrangler/src/dev.tsx:
```diff
@@ -391,13 +397,36 @@ function useCustomBuild(
+        (_event, _path) => {
+          console.log("Restarting build...");
+          cmd.kill();
+          cmd = execa(commandPieces[0], commandPieces.slice(1), {
+            ...(cwd && { cwd }),
+            stderr: "inherit",
+            stdout: "inherit",
+          });
```
This block is now repeated - let's move it into its own function?

threepointone replied:
It's just one statement, doesn't pass my threshold for a new abstraction yet imo

threepointone commented on package.json:
```diff
@@ -30,7 +30,8 @@
     "check": "prettier packages/** --check && tsc && npm run lint",
-    "prettify": "prettier packages/** --write"
+    "prettify": "prettier packages/** --write",
+    "test": "npm run test --workspace=wrangler"
```
Strongly agreed, just did this as a quick dx win for myself rn.
